File: core/src/com/nopalsoft/thetruecolor/scene2d/DialogHelpSettings.java
package com.nopalsoft.thetruecolor.scene2d;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.ui.ButtonGroup;
import com.badlogic.gdx.scenes.scene2d.ui.Image;
import com.badlogic.gdx.scenes.scene2d.ui.Label;
import com.badlogic.gdx.scenes.scene2d.ui.Label.LabelStyle;
import com.badlogic.gdx.scenes.scene2d.ui.Table;
import com.badlogic.gdx.scenes.scene2d.ui.TextButton;
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener;
import com.badlogic.gdx.scenes.scene2d.utils.TextureRegionDrawable;
import com.nopalsoft.thetruecolor.Assets;
import com.nopalsoft.thetruecolor.Settings;
import com.nopalsoft.thetruecolor.screens.Screens;

public class DialogHelpSettings extends Dialog {
    static final float WIDTH = 440;
    static final float HEIGHT = 600;

    public enum Languages {
        DEFAULT, ENGLISH, SPANISH, CHINESE_TAIWAN, RUSSIAN, FRENCH, JAPANESE, PORTUGUESE
    }

    Table tbColores;

    TextButton btDefault, btEnglish, btSpanish, btChineseTaiwan, btRussian, btFrench, btJapanese, btPortugese;
    TextButton btMore;

    DialogMoreLanguages vtnaMoreLanguages;

    public DialogHelpSettings(final Screens currentScreen) {
        super(currentScreen, WIDTH, HEIGHT, 80);
        setCloseButton(560);

        Label lbIdioma = new Label(Assets.languages.get("language"), new LabelStyle(Assets.fontSmall, Color.BLACK));
        lbIdioma.setPosition(getWidth() / 2f - lbIdioma.getWidth() / 2f, 555);
        addActor(lbIdioma);

        vtnaMoreLanguages = new DialogMoreLanguages(currentScreen);

        btMore = crearBotton(Assets.languages.get("more"), null, Assets.flagMore);
        btMore.addListener(new ClickListener() {
            @Override
            public void clicked(InputEvent event, float x, float y) {
                btMore.setChecked(false);
                vtnaMoreLanguages.show(currentScreen.stage);
            }
        });

        btDefault = crearBotton("Default", Languages.DEFAULT, null);

        btEnglish = crearBotton("English", Languages.ENGLISH, Assets.flagEnglish);

        btSpanish = crearBotton("Espalier", Languages.SPANISH, Assets.flagSpanish);

        btChineseTaiwan = crearBotton("中文", Languages.CHINESE_TAIWAN, Assets.flagChinese_TW);

        btRussian = crearBotton("Русский", Languages.RUSSIAN, Assets.flagRussian);
        btRussian.getLabel().setFontScale(.7f);

        btFrench = crearBotton("Français", Languages.FRENCH, Assets.flagFrench);

        btJapanese = crearBotton("日本語", Languages.JAPANESE, Assets.flagJapanese);

        btPortugese = crearBotton("Português", Languages.PORTUGUESE, Assets.flagPortugese);

        switch (Settings.selectedLanguage) {
            case DEFAULT:
                btDefault.setChecked(true);
                break;
            case ENGLISH:
                btEnglish.setChecked(true);
                break;
            case SPANISH:
                btSpanish.setChecked(true);
                break;
            case CHINESE_TAIWAN:
                btChineseTaiwan.setChecked(true);
                break;
            case RUSSIAN:
                btRussian.setChecked(true);
                break;
            case FRENCH:
                btFrench.setChecked(true);
                break;
            case JAPANESE:
                btJapanese.setChecked(true);
                break;
            case PORTUGUESE:
                btPortugese.setChecked(true);
                break;
        }


        ButtonGroup<TextButton> btGroup = new ButtonGroup<>(btDefault, btEnglish, btSpanish, btChineseTaiwan, btRussian, btFrench, btJapanese, btPortugese);
        btGroup.setMaxCheckCount(1);

        Table tbIdiomas = new Table();
        tbIdiomas.setSize(getWidth(), 200);
        tbIdiomas.setPosition(0, 300);

        tbIdiomas.defaults().expandX().pad(3f, 10, 3f, 10).fill().uniform();

        tbIdiomas.add(btDefault);
        tbIdiomas.add(btEnglish);
        tbIdiomas.add(btSpanish);
        tbIdiomas.row();
        tbIdiomas.add(btChineseTaiwan);
        tbIdiomas.add(btRussian);
        tbIdiomas.add(btFrench);
        tbIdiomas.row();
        tbIdiomas.add(btJapanese);
        tbIdiomas.add(btPortugese);
        tbIdiomas.add(btMore);
        tbIdiomas.row();

        // Los colores
        tbColores = new Table();
        tbColores.setSize(getWidth(), 240);

        fillTableColores();

        addActor(tbColores);
        addActor(tbIdiomas);
    }

    private void fillTableColores() {
        tbColores.clear();

        Image iBlue = new Image(Assets.barTimer);
        iBlue.setColor(Color.BLUE);

        Image iCyan = new Image(Assets.barTimer);
        iCyan.setColor(Color.CYAN);

        Image iGreen = new Image(Assets.barTimer);
        iGreen.setColor(Color.GREEN);

        Image iYellow = new Image(Assets.barTimer);
        iYellow.setColor(Color.YELLOW);

        Image iPink = new Image(Assets.barTimer);
        iPink.setColor(Color.PINK);

        Image iBrown = new Image(Assets.barTimer);
        iBrown.setColor(new Color(.6f, .3f, 0, 1));

        Image iPurple = new Image(Assets.barTimer);
        iPurple.setColor(Color.PURPLE);

        Image iRed = new Image(Assets.barTimer);
        iRed.setColor(Color.RED);

        tbColores.defaults().expandX().padTop(5).padBottom(5);

        tbColores.add(getNewLabelWithColor(languages.get("blue"), Color.BLUE));
        tbColores.add(iBlue).size(40).left();

        tbColores.add(getNewLabelWithColor(languages.get("cyan"), Color.CYAN));
        tbColores.add(iCyan).size(40).left();

        tbColores.row();
        tbColores.add(getNewLabelWithColor(languages.get("green"), Color.GREEN));
        tbColores.add(iGreen).size(40).left();

        tbColores.add(getNewLabelWithColor(languages.get("yellow"), Color.YELLOW));
        tbColores.add(iYellow).size(40).left();

        tbColores.row();
        tbColores.add(getNewLabelWithColor(languages.get("pink"), Color.PINK));
        tbColores.add(iPink).size(40).left();

        tbColores.add(getNewLabelWithColor(languages.get("brown"), new Color(.6f, .3f, 0, 1)));
        tbColores.add(iBrown).size(40).left();

        tbColores.row();
        tbColores.add(getNewLabelWithColor(languages.get("purple"), Color.PURPLE));
        tbColores.add(iPurple).size(40).left();

        tbColores.add(getNewLabelWithColor(languages.get("red"), Color.RED));
        tbColores.add(iRed).size(40).left();
    }

    private Label getNewLabelWithColor(String text, Color color) {
        LabelStyle lbStyleColores = new LabelStyle(Assets.fontSmall, color);
        Label label = new Label(text, lbStyleColores);
        if (Settings.selectedLanguage == Languages.RUSSIAN) {
            label.setFontScale(.7f);
        }
        return label;

    }

    private TextButton crearBotton(String texto, Languages language, TextureRegionDrawable flag) {
        TextButton btAux = new TextButton(texto, Assets.textButtonStyle);
        if (flag != null) {
            btAux.add(new Image(flag));
        }
        if (language != null) {
            btAux.addListener(addClickListener(language));
        }
        btAux.getLabel().setFontScale(.8f);
        return btAux;

    }

    private ClickListener addClickListener(final Languages language) {
        return new ClickListener() {
            @Override
            public void clicked(InputEvent event, float x, float y) {
                Settings.selectedLanguage = language;
                Settings.save();
                Assets.loadAssetsWithSettings();
                languages = Assets.languages;
                fillTableColores();

            }
        };
    }

}

File: core/src/com/nopalsoft/thetruecolor/scene2d/DialogHelpSettings.kt
package com.nopalsoft.thetruecolor.scene2d

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.scenes.scene2d.InputEvent
import com.badlogic.gdx.scenes.scene2d.ui.ButtonGroup
import com.badlogic.gdx.scenes.scene2d.ui.Image
import com.badlogic.gdx.scenes.scene2d.ui.Label
import com.badlogic.gdx.scenes.scene2d.ui.Label.LabelStyle
import com.badlogic.gdx.scenes.scene2d.ui.Table
import com.badlogic.gdx.scenes.scene2d.ui.TextButton
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener
import com.badlogic.gdx.scenes.scene2d.utils.TextureRegionDrawable
import com.nopalsoft.thetruecolor.Assets
import com.nopalsoft.thetruecolor.Assets.loadAssetsWithSettings
import com.nopalsoft.thetruecolor.Settings
import com.nopalsoft.thetruecolor.Settings.save
import com.nopalsoft.thetruecolor.screens.BaseScreen

class DialogHelpSettings(currentScreen: BaseScreen) :
    BaseDialog(currentScreen, WIDTH, HEIGHT, 80F) {

    private var tableColors = Table()

    private val buttonDefault = createButton("Default", Languages.DEFAULT, null)
    private val buttonEnglish = createButton("English", Languages.ENGLISH, Assets.flagEnglish)
    private val buttonSpanish = createButton("Espalier", Languages.SPANISH, Assets.flagSpanish)
    private val buttonChineseTaiwan = createButton("中文", Languages.CHINESE_TAIWAN, Assets.flagChinese_TW)
    private val buttonRussian = createButton("Русский", Languages.RUSSIAN, Assets.flagRussian)
    private val buttonFrench = createButton("Français", Languages.FRENCH, Assets.flagFrench)
    private val buttonJapanese = createButton("日本語", Languages.JAPANESE, Assets.flagJapanese)
    private val buttonPortugese = createButton("Portuguese", Languages.PORTUGUESE, Assets.flagPortugese)
    private val buttonMore = createButton(languages.get("more"), null, Assets.flagMore)

    private val dialogMoreLanguages = DialogMoreLanguages(currentScreen)


    init {
        setCloseButton(560f)
        val labelLanguage = Label(languages.get("language"), LabelStyle(Assets.fontSmall, Color.BLACK))
        labelLanguage.setPosition(width / 2f - labelLanguage.width / 2f, 555f)
        addActor(labelLanguage)

        buttonMore.addListener(object : ClickListener() {
            override fun clicked(event: InputEvent, x: Float, y: Float) {
                buttonMore.isChecked = false
                dialogMoreLanguages.show(currentScreen.stage)
            }
        })
        buttonRussian.label.setFontScale(.7f)

        when (Settings.selectedLanguage) {
            Languages.DEFAULT -> buttonDefault.isChecked = true
            Languages.ENGLISH -> buttonEnglish.isChecked = true
            Languages.SPANISH -> buttonSpanish.isChecked = true
            Languages.CHINESE_TAIWAN -> buttonChineseTaiwan.isChecked = true
            Languages.RUSSIAN -> buttonRussian.isChecked = true
            Languages.FRENCH -> buttonFrench.isChecked = true
            Languages.JAPANESE -> buttonJapanese.isChecked = true
            Languages.PORTUGUESE -> buttonPortugese.isChecked = true
        }

        val btGroup = ButtonGroup(
            buttonDefault,
            buttonEnglish,
            buttonSpanish,
            buttonChineseTaiwan,
            buttonRussian,
            buttonFrench,
            buttonJapanese,
            buttonPortugese
        )
        btGroup.setMaxCheckCount(1)

        val tableLanguages = Table()
        tableLanguages.setSize(width, 200f)
        tableLanguages.setPosition(0f, 300f)

        tableLanguages.defaults().expandX().pad(3f, 10f, 3f, 10f).fill().uniform()

        tableLanguages.add(buttonDefault)
        tableLanguages.add(buttonEnglish)
        tableLanguages.add(buttonSpanish)
        tableLanguages.row()
        tableLanguages.add(buttonChineseTaiwan)
        tableLanguages.add(buttonRussian)
        tableLanguages.add(buttonFrench)
        tableLanguages.row()
        tableLanguages.add(buttonJapanese)
        tableLanguages.add(buttonPortugese)
        tableLanguages.add(buttonMore)
        tableLanguages.row()

        // The colors
        tableColors = Table()
        tableColors.setSize(width, 240f)

        fillTableColors()

        addActor(tableColors)
        addActor(tableLanguages)
    }

    private fun createButton(
        text: String,
        language: Languages?,
        flag: TextureRegionDrawable?
    ): TextButton {
        val textButton = TextButton(text, Assets.textButtonStyle)
        if (flag != null) {
            textButton.add(Image(flag))
        }
        if (language != null) {
            textButton.addListener(addClickListener(language))
        }
        textButton.label.setFontScale(.8f)
        return textButton
    }

    private fun addClickListener(language: Languages): ClickListener {
        return object : ClickListener() {
            override fun clicked(event: InputEvent, x: Float, y: Float) {
                Settings.selectedLanguage = language
                save()
                loadAssetsWithSettings()
                fillTableColors()
            }
        }
    }

    private fun fillTableColors() {
        tableColors.clear()

        val iBlue = Image(Assets.barTimer)
        iBlue.color = Color.BLUE

        val iCyan = Image(Assets.barTimer)
        iCyan.color = Color.CYAN

        val iGreen = Image(Assets.barTimer)
        iGreen.color = Color.GREEN

        val iYellow = Image(Assets.barTimer)
        iYellow.color = Color.YELLOW

        val iPink = Image(Assets.barTimer)
        iPink.color = Color.PINK

        val iBrown = Image(Assets.barTimer)
        iBrown.color = Color(.6f, .3f, 0f, 1f)

        val iPurple = Image(Assets.barTimer)
        iPurple.color = Color.PURPLE

        val iRed = Image(Assets.barTimer)
        iRed.color = Color.RED

        tableColors.defaults().expandX().padTop(5f).padBottom(5f)

        tableColors.add(getNewLabelWithColor(languages["blue"], Color.BLUE))
        tableColors.add(iBlue).size(40f).left()

        tableColors.add(getNewLabelWithColor(languages["cyan"], Color.CYAN))
        tableColors.add(iCyan).size(40f).left()

        tableColors.row()
        tableColors.add(getNewLabelWithColor(languages["green"], Color.GREEN))
        tableColors.add(iGreen).size(40f).left()

        tableColors.add(getNewLabelWithColor(languages["yellow"], Color.YELLOW))
        tableColors.add(iYellow).size(40f).left()

        tableColors.row()
        tableColors.add(getNewLabelWithColor(languages["pink"], Color.PINK))
        tableColors.add(iPink).size(40f).left()

        tableColors.add(getNewLabelWithColor(languages["brown"], Color(.6f, .3f, 0f, 1f)))
        tableColors.add(iBrown).size(40f).left()

        tableColors.row()
        tableColors.add(getNewLabelWithColor(languages["purple"], Color.PURPLE))
        tableColors.add(iPurple).size(40f).left()

        tableColors.add(getNewLabelWithColor(languages["red"], Color.RED))
        tableColors.add(iRed).size(40f).left()
    }

    private fun getNewLabelWithColor(text: String, color: Color): Label {
        val labelStyleColors = LabelStyle(Assets.fontSmall, color)
        val label = Label(text, labelStyleColors)
        if (Settings.selectedLanguage == Languages.RUSSIAN) {
            label.setFontScale(.7f)
        }
        return label
    }

    enum class Languages {
        DEFAULT, ENGLISH, SPANISH, CHINESE_TAIWAN, RUSSIAN, FRENCH, JAPANESE, PORTUGUESE
    }

    companion object {
        private const val WIDTH = 440F
        private const val HEIGHT = 600F

    }
}